lib/puppet/util/windows/adsi.rb

```diff
@@ -176,8 +191,16 @@ def exists?(name_or_sid)
         well_known = false
         if (sid = Puppet::Util::Windows::SID.name_to_principal(name_or_sid))
           # Examples of SidType include SidTypeUser, SidTypeGroup
-          return true if sid.account_type == "SidType#{@object_class.capitalize}".to_sym
-  
+          if sid.account_type == "SidType#{@object_class.capitalize}".to_sym
```

Comment:
Need to test that this doesn't contradict `list_all` i.e.  `list_all` should only return local users/groups.  This was the usecase in the original ticket.

This needs to be tested for both Users and Groups as this commit is affecting both.

Reply:
not sure I parse...

* `list_all` will definitely only return local users/groups (as it uses WMI to query them), the issue in this ticket stems from the use of LookupAccountNameW which falls back to domain
* I manually tested this with both users and groups and `ensure => absent` is idempotent (first run it removes the local user/group, second run it does nothing - even though the user/group exists on the domain)

Reply:
My comment has mainly about:

* list_all uses the WMI query to find resources

* wheres `exists?` uses the LookupSIDAccountW API call.

The expectation would be that
* Anything IN `list_all` should return `true` for `exists?`
* Anything NOT IN `list_all` should return `false` for `exists?`

I did a quick test on the Puppet 5.x branch and it seems that this is indeed actually broken.  i.e.

* Given a group called `domaingroup` in an AD Domain
  * The name DOES not appear in `list_all`
  * Calling `...exists?('domaingroup')` raises instead of returning false.

So this is a 💯  👍 for me for this PR.  Fixes an obscure behaviour.